libs/s25main/mapGenerator/HeadQuarters.h

```diff
@@ -42,45 +44,66 @@ namespace rttr { namespace mapGenerator {
     {
         if(area.empty())
         {
             throw std::runtime_error("Could not find any valid HQ position!");
         }
 
-        const auto isValid = [](const MapPoint& pt) { return pt.isValid(); };
-        const auto isObstacle = [&map](MapPoint point) {
-            return map.textureMap.Any(point, [](auto t) { return !t.Is(ETerrain::Buildable); });
+        const auto isObstacle = [&map](const MapPoint& pt) {
+            return map.textureMap.Any(pt, [](auto t) { return !t.Is(ETerrain::Buildable); });
         };
-        std::vector<MapPoint> hqs;
-        std::copy_if(map.hqPositions.begin(), map.hqPositions.end(), std::back_inserter(hqs), isValid);
 
-        // Quality of a map point as one player's HQ is a mix of:
-        // 1. distance to other players' HQs (higher = better)
-        // 2. distance to any other obstacle e.g. mountain & water (higher = better)
-        NodeMapBase<unsigned> potentialHqQuality = DistancesTo(hqs, map.size);
+        // Ensure possible HQ positions are far enough from any obstacles
         const auto obstacleDistance = DistancesTo(map.size, isObstacle);
-        const auto minDistance = helpers::clamp(GetMaximum(obstacleDistance, area), 2u, 4u);
+        const auto minObstacleDistance = helpers::clamp(GetMaximum(obstacleDistance, area), 2u, 4u);
+        const auto farFromObstacles = [&obstacleDistance, minObstacleDistance](const MapPoint& pt) {
+            return obstacleDistance[pt] >= minObstacleDistance;
+        };
+        std::vector<MapPoint> possiblePositions;
+        std::copy_if(area.begin(), area.end(), std::back_inserter(possiblePositions), farFromObstacles);
 
+        if(possiblePositions.empty())
+        {
+            return possiblePositions;
+        }
+
+        // Keep minimum distance to other HQs and desired distance to mountains
+        const auto mountain = [&map](const MapPoint& pt) { return map.textureMap.Any(pt, IsMinableMountain); };
+        const auto mountainDistances = DistancesTo(map.size, mountain);
+        const auto desiredDistance = static_cast<unsigned>(distance);
+        const auto distanceToOtherHqs = DistancesTo(map.hqPositions, map.size);
+        const unsigned minimumHqDistance = (map.size.x + map.size.y) / 16;
+        const int maximumRadius = (map.size.x + map.size.y) / 4;
         std::vector<MapPoint> positions;
-        for(const MapPoint& pt : area)
+        for(int radius = 1; positions.empty() && radius < maximumRadius; radius *= 2)
         {
-            if(obstacleDistance[pt] >= minDistance)
-            {
-                positions.push_back(pt);
-            } else
+            for(const MapPoint& pt : possiblePositions)
             {
-                potentialHqQuality[pt] = 0;
+                if(std::abs(static_cast<int>(mountainDistances[pt] - desiredDistance)) < radius
+                   && distanceToOtherHqs[pt] > minimumHqDistance)
```

Comment:
Makes sense. Then 2 new questions:

1. Is it intentional that the minimumHqDistance is ignored in the fallback below?
2. Would it make sense to remove those non-matching points first, before the radius-loop? If 1. is not intentional, then I'd certainly do so, otherwise it might still be a good idea but is not that crucial.

Also the cast is misplaced: If `desiredDistance > mountainDistances[pt]` then you get UB because the value is very large and hence not representable in int. Only the int->unsigned conversion is valid. I know this is fine in practice but I wouldn't want to introduce UB here.

And I still think the algorithm is not the right choice here, yes it works and is simple enough but the worst case behavior could be bad. See my previous idea about sorting `possiblePositions`. Seeing the current change I'd even go further: Sort by the difference (the `std::abs`-result) and then simply take the first points with the same radius. Sorting could also be done in a vector of `pair<MapPoint, distanceDifference>`.
Or, maybe still better: It is possible in a single pass:

- Start with `minDistanceDifference=<maxint>`
- Go exactly once over possiblePositions skipping invalid ones (minimumHqDistance)
- If that points distanceDifference is equal to minDistanceDifference add to possiblePoints
- if it is lower, clear possiblePoints, reset minDistanceDifference and add to possiblePoints

Because that's what you are trying here, aren't you: Selecting the points with the least difference of mountainDistances to desiredDistance

Then the only way that `positions` is empty afterwards is for the case when all points are less than minimumHqDistance (BTW: If it's a minimum I'd expect that this distance is valid, hence `distanceToOtherHqs[pt] >= minimumHqDistance`, oversight or intentional?)

To me this sounds good, what do you think? Am I missing anything or is there a advantage of the current implementation?

Reply:
I think you're right, I'll take a look at this probably tomorrow!

Reply:
Just updated the PR. To answer your question:
> Is it intentional that the minimumHqDistance is ignored in the fallback below?
Yes, that was intentional. In case we can't find any suitable position with mountains in range & minimum HQ distance, then we just take the position with most distance to other HQs.